### ISSUE TITLE
About a merger of `sequelize-typescript` with `sequelize`

### ISSUE THREAD

**Issue by ephys** [sequelize/sequelize-typescript#1321]
Hi again 👋

At Sequelize, we consider that `sequelize-typescript` has proven itself as being the best way of using Sequelize with TypeScript. We actually [recommend it officially in our documentation](https://sequelize.org/docs/v6/other-topics/typescript/).

We're working on [improving our TypeScript support](https://github.com/sequelize/sequelize/issues?q=is%3Aopen+is%3Aissue+label%3ARFC+label%3A%22type%3A+typescript%22) in core. We have various RFC but we're thinking merging with `sequelize-typescript` instead of implementing our own APIs would ease the transition and be a better option for our users.

After some initial discussion with @Divlo, it looks like there may be interest on your end too. Unfortunately I couldn't figure out how to contact the other maintainers

What would be the next steps?

## Team Merge

If you are currently a maintainer of `sequelize-typescript`, we would like to invite you to join the `sequelize` team. Entirely up to you, of course :)

## Code Merge

Here is a breakdown of the features we would like to move inside core:

- Dynamic loading of models through something like `sequelize.addModels(globOrModelArray)`: Yes
- Decorators: Absolutely
- Type-safe auto generated functions (`$get`, `$set`): Likely not, we are working on a different API for this: https://github.com/sequelize/sequelize/issues/14302. Maybe we can bring it over, and deprecate it once the new API is released, to ease transition?
- Repositories: Maybe, undecided yet

## Repository & npm transfer?

Is transfering the project over to our npm org, and transfering the repository over to our github org, something you would consider?

Many issues are actually caused by Sequelize, but we can't close them or transfer them to the right repository. Once v6 is EOL, we would also like to deprecate the npm package as everything will be in core.

**Comment by RobinBuschmann** [comment]
Hey @ephys, thanks for reaching out. 
> If you are currently a maintainer of sequelize-typescript, we would like to invite you to join the sequelize team. Entirely up to you, of course :)

Unfortunately I still have no time to maintain sequelize-typescript anymore. Especially due to private matters.

> Is transfering the project over to our npm org, and transfering the repository over to our github org, something you would consider?

Sounds great to me too. Since @Divlo and @lukashroch reacted with 👍 I assume they are ok if I transfer this repo as soon as possible. But from my understanding it is not necessary to transfer the npm package, but to just add maintainers. Am I correct with this? The npm package sequelize isn't owned by an organization as well. The ownership of sequelize belongs to a private person, right?

**Comment by lukashroch** [comment]
Sounds good to me ;-)

Repo can be easily transferred to org on Github. But I think in case of npm package, you will have deprecate original package under non-scoped name (set up deprecation message with new link) and publish under new org @scope.

In terms of code merging, integration of majority (there are already some bits deprecated so these could go away with merger) of the code with same API in first step would be best for users to migrate as much painlessly as possible. And then start introducing new API and deprecated old API (like `$set` etc) gradually.

**Comment by WikiRik** [comment]
> But from my understanding it is not necessary to transfer the npm package, but to just add maintainers. Am I correct with this? The npm package sequelize isn't owned by an organization as well. The ownership of sequelize belongs to a private person, right?

That is correct, but starting with v7 we are releasing it under the sequelize organisation. For sequelize-typescript it is enough to add any combination of @ephys @sdepold and me.

Thanks for the response!

**Comment by sdepold** [comment]
This certainly is exciting :) happy to hear that you are open for the conversation and action in general

**Comment by ephys** [comment]
> But from my understanding it is not necessary to transfer the npm package, but to just add maintainers.

You're right :)

> you will have deprecate original package under non-scoped name

I think what we should continue to use the `sequelize-typescript` package in v6.
In v7, I would propose to add it directly in core: `@sequelize/core/decorators-legacy` (legacy because we'll need to add support for stage 3 decorators soon).
`sequelize-typescript` would then have the same EOL date as Sequelize 6, and we'll deprecate it once Sequelize 6 is deprecated :)

I also agree that adding everything that is not already deprecated, then deprecate things as we introduce the new APIs, is the way to go

To sum up our next points of action:

- Transfer repository to the sequelize org
- Add us to the npm project for future releases (I'm also `ephys` on npm)
- Open a PR to merge the two codebases (I think I will handle this part, to get familiar with the codebase)

**Comment by RobinBuschmann** [comment]
When I try to transfer sequelize-typscript to the organisation sequelize, github complaints about my permissions as follows:
> You don’t have the permission to create public repositories on sequelize

So I assume before I can do so, I need to be a member of the sequelize organisation with the required rights.

**Comment by RobinBuschmann** [comment]
@ephys @sdepold As you've already noticed, I invited you to st on npm

**Comment by ephys** [comment]
I have, thank you :)
I'll take a look at what's going on with the permission

**Comment by sdepold** [comment]
> When I try to transfer sequelize-typscript to the organisation sequelize, github complaints about my permissions as follows:
> 
> > You don’t have the permission to create public repositories on sequelize
> 
> 
> 
> So I assume before I can do so, I need to be a member of the sequelize organisation with the required rights.

I think you should find an invitation to join the team on GitHub now and if we are lucky that will suffice to create and transfer repos. If not maybe it would be the easiest if you at least temporarily join our slack :)

**Comment by RobinBuschmann** [comment]
@sdepold  Thanks! 

I've joined sequelize, but when trying to transfer st I still see:
>You don’t have the permission to create public repositories on sequelize

The docs just says the same:
>To transfer a repository that you own to an organization, you must have permission to create a repository in the target organization.

So it seems so, that I'm missing that permission.

**Comment by sdepold** [comment]
@RobinBuschmann please try again :)

**Comment by sdepold** [comment]
Seems that it worked out, right? Just made you a member of Sequelize again (as opposed to owner). Hope that works for you :)

**Comment by sdepold** [comment]
Thanks a bunch! :)

**Comment by RobinBuschmann** [comment]
Thank you for bringing this up :) I think this is the right way to go for sequelize-typescript or its features

> Seems that it worked out, right? Just made you a member of Sequelize again (as opposed to owner). Hope that works for you :)
Yes, this will work for me.

**Comment by RobinBuschmann** [comment]
Closing this, because sequelize-typescript is now part of the sequelize org

**Comment by ephys** [comment]
Thank you! :)
